# 텐서플로우 기초

https://www.tensorflow.org/guide/effective_tf2?hl=ko

#### tensorflow 설치

https://www.tensorflow.org/install/pip?hl=ko

**[pip 에서 설치]**

tensorflow 와 Cuda 를 설치

```sh
$ pip install --upgrade pip
$ pip install tensorflow[and-cuda]
$ pip install keras --upgrade
```


**[conda 로  설치]**

```sh
conda install -c conda-forge  tensorflow[and-cuda]
conda install -c conda-forge keras
```


scikit-learn 도 유틸리티로 자주 사용된다.

```sh
$ pip install -U scikit-learn
```


텐서플로우 설치 확인

In [1]:
import tensorflow as tf
import numpy as np

print(tf.__version__, np.__version__)

2025-07-22 07:11:19.183615: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-07-22 07:11:19.617300: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1753135879.780391  168080 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1753135879.827746  168080 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1753135880.174145  168080 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking 

2.19.0 2.1.3


In [2]:
import tensorflow as tf; print(tf.reduce_sum(tf.random.normal([1000, 1000])))

tf.Tensor(83.12167, shape=(), dtype=float32)


I0000 00:00:1753135889.767762  168080 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 825 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4060 Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.9


텐서플로우 기본

TensorFlow의 가장 기본적인 빌딩 블록인 Tensor, Operation, 그리고 Variable에 대해 살펴보고, 이 개념들을 확실히 이해하는 것이 TensorFlow를 이용한 딥러닝 모델 구축의 시작이다.


1. 텐서
1. 상수
1. Operation(연산)
1. 변수
1. 행렬

> Test
> - tensorflow 2.19, keras 3.10
> - tensorflow 2.9
> - tensorflow 2.7

In [3]:
import numpy as np
import tensorflow as tf
import keras
tf.__version__, keras.__version__

('2.19.0', '3.10.0')

## GPU

In [4]:
# tf.test.is_gpu_available()
tf.config.list_physical_devices('GPU')

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [5]:
[device.name for device in tf.config.experimental.list_physical_devices()]

['/physical_device:CPU:0', '/physical_device:GPU:0']

# 1. AI의 데이터 형식



## -Tensor

AI 분야에서 데이터를 표현하는 기본 단위로 텐서를 사용한다. 스칼라(0차원), 벡터(1차원), 행렬(2차원)을 포함하여 그 이상 고차원의 데이터를 모두 텐서라고 부르고 AIdptj *텐서* 를 기본 데이터 구조로 사용한다. 

텐서(tensor)는 데이터를 위한 컨테이너, 거의 수치형 데이터를 다루므로 숫자를 위한 컨테이너이다. 텐서는 임의 차원 개수를 가진 행렬로 행렬은 아래 같은 종류가 있다.

 - 0차원 텐서: Scalar, 0D 텐서.
    * 나이, 온도, 가격
 - 1차원 텐서: Vector 벡터, 1D텐서.
    * 열을 축으로하는 1차원 배열
    * `[1, 3, 9]`
 - 2차원 텐서: Matrix 행렬, 2D 텐서. 
    * 행과 열을 축으로하는 2차원.
    * 8x8 크기 흑백 이미지 데이터
 - 3차원 텐서: 행렬의 배열 의미. 
    * 깊이 또는 채널이라는 3번째 방향을 갖는다.
    * 8x8x3 크기의 컬러 이미지
 - 4차원 텐서: 3차원 텐서의 배열
    * 동영상 데이터: 보통 (샘플수x프레임수x높이x너비x채널)로 표현 
    * 배치 데이터: 128개 이미지를 1배치로 처리시 (128, 256, 256, 3) 형태

<img src='https://i.imgur.com/PRAiV7j.png' width=600>

## -TensorFlow의 Tensor

Tensorflow의 텐서는 NumPy 배열과 매우 유사하지만, GPU 가속을 지원하고 딥러닝 연산에 최적화되어 있다는 차이점이 있다.

> TensorFlow 2.x에서는 텐서 연산에 **Eager Execution(즉시 실행)** 이 기본으로 설정되어 있어, 코드를 작성하는 즉시 연산이 실행되고 결과가 반환됩니다. 이는 Python의 일반적인 코딩 방식과 유사하여 디버깅 및 개발이 훨씬 용이합니다.

## -텐서의 구성 요소:

- 랭크 (Rank) 또는 차원 (Dimension)
  - 텐서의 차원 수를 나타냅니다. 예를 들어, 스칼라는 0차원, 벡터는 1차원, 행렬은 2차원입니다.
- 형태 (Shape)
  -  각 차원의 크기를 나타내는 튜플입니다. 예를 들어, 2x3 행렬은 (2, 3) 형태를 가집니다.
- 데이터 타입 (Dtype)
  -  텐서가 저장하는 요소의 데이터 타입입니다. 정수(tf.int32, tf.int64), 부동 소수점(tf.float32, tf.float64), 불리언(tf.bool) 등 다양한 타입을 지원합니다. 
  - 연산 효율성을 위해 보통 딥러닝에서는 tf.float32를 많이 사용합니다.

## - 텐서 샘플

 - 0D 텐서
    * 스칼라 값. 예) 넓이 100cm
 - 1D 텐서
    * 벡터(배열) 데이터. 
    * 예) 학생 명단 [홍길동,남원길,서울역]
 - 2D 텐서
    * 행렬
    - 예) (samples, feature)
 - 3D 텐서
    * 행렬의 배열. 시계열, 
    * 시퀀스 데이터 (samples, timsteps, feature)
    * 256x256 픽셀 RGB 컬러 이미지 (256, 256, 3)
 - 4D 텐서
    * 채널 이미지 데이터\
     (samples, height, width, channels), \
     (samples, channels, height, width)
 - 동영상 데이터
    * 4D 이상 고차원 텐서로 4D의 배열
    * (samples, frames, height, width, channels),\
     (samples, frames, channels, height, width) 
    * 예: 10개의 비디오 클립, 각 클립은 30프레임, 128x128 픽셀, RGB 색상을 가진 경우:\
     (10, 30, 128, 128, 3)


<img src='https://i.imgur.com/Ehn05pP.png' width=700>

### 동영상 텐서의 Channel First & Last

TensorFlow에서 이미지 데이터 처리 시 흔히 사용되는 channels_last 형식을 따른다
  - 비디오 텐서 (samples, frames, height, width, channels) 형식
  - 단일 이미지 텐서는 ( height, width, channels) 형태


PyTorch의 경우, 이미지 데이터 처리에서 channels_first 형식을 선호하는 경향
 - 단일 이미지 텐서는 (channels, height, width) 형태
 - 비디오 텐서 (samples, channels, frames, height, width) \
   또는 (samples, frames, channels, height, width)가 사용.


#### 스칼라 (0D tensor)

하나의 숫자 값을 담고 있는 텐서로서 0차원, 0D텐서이다. 0차원 텐서의 축의 개수는 0으로 축(ndim)을 랭크(rank)라고도 부른다.

In [6]:
# 스칼라
x = np.array(1)
x

array(1)

In [7]:
x.shape, x.ndim, x

((), 0, array(1))

#### 벡터 (1D Tensor)

숫자 배열을 벡터, 1D 텐서라고 부른다. 1D 텐서는 하나의 축을 가진다.

In [8]:
# 벡터
x = np.array([1,2,3,4])
x.shape, x.ndim, x

((4,), 1, array([1, 2, 3, 4]))

위 벡터는 4개의 요소를 가지고 있어서 5차원 벡터라고 부른다. ( 4D 텐서와 혼동하지 말것!)

#### 행렬 (2D Tensor)

행렬은 2차원으로 구성되어 2개의 축을 가진다.
- (samples, feature) 형식으로 구성된다.
  - 10만명의 나이/우편번호/소득 으로 구성된 인구 데이터: (100000, 3) 크기 텐서
  - 텍스트에 공통단어 2만개에서 단어의 등장횟수로 표현된 경우: (500, 20000) 

In [9]:
# 행렬
x = np.array([[1,2,3],[4,5,6],[7,8,9]])
x.shape, x.ndim, x

((3, 3),
 2,
 array([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]]))

#### 3차원 배열 ( 3D Tensor)

배열의 3차원 축을 가진 배열을 3D Tensor라고 한다.

- 시계열/시퀀스 데이터
  - 데이터의 시간/순서가 중요하면 시간축을 포함해 (samples, timsteps, feature) 크기 3차원 텐서로 구성된다.
  - 주식 가격 데이터: 1분마다 주식가격, 1분마다 최고/최소 가격. 
      - 하루 거래 (390,3)크기 2D텐서
      - 연간 250일 (250, 390, 3) 크기 3D텐서
  - 트윗 데이터 세트: 트윗은 128개 알파벳, 280 문자로. 128개 2진벡터로 구성( 280, 128)
      - 100만개 트쉿은 (1000000, 280, 128) 텐서로 구성

In [10]:
# 텐서: 3d
x = np.array([[[1,2,3,4,5],
              [6,7,8,9,10]],
             [[0,-1,-2,-3,-4],
             [-5,-6,-7,-8,-9]]])
x.shape, x.ndim, x

((2, 2, 5),
 3,
 array([[[ 1,  2,  3,  4,  5],
         [ 6,  7,  8,  9, 10]],
 
        [[ 0, -1, -2, -3, -4],
         [-5, -6, -7, -8, -9]]]))

####  4D 텐서

이미지는 높이/너비/컬러의 3차원. 흑백도 높이/너비/컬러.

 - 128개 256x256 크기 흑백 배치: (128, 256, 256, 1) 
 - 128개 256x256 크기 컬러 배치: (128, 256, 256, 3)
 - 이미지 텐서 크지 지정은 채널 마지막(channel-last) 와 채널 우선(channel-first)
    - Theano 는 (128, 1, 256, 256)
    - Keras 는 모두 지원



In [12]:
img_tensor = (128, 256, 256, 4)
img = np.random.rand(*img_tensor)

In [13]:
img.shape

(128, 256, 256, 4)

####  5D 텐서

동영상 데이터 비디오는 프레임의 연속. 프레임원 컬러 이미지, 현실에서 5D 차원 가진 몇 안되는 데이터




In [14]:
tensor_shape = (10, 30, 128, 128, 3)
mov = np.random.rand(*tensor_shape)
mov.shape

(10, 30, 128, 128, 3)

In [15]:
mov[0]

array([[[[0.97822117, 0.27696432, 0.31065699],
         [0.16691156, 0.96019574, 0.35712579],
         [0.52019033, 0.1302745 , 0.07639221],
         ...,
         [0.32006658, 0.45160375, 0.24743383],
         [0.42856348, 0.22895247, 0.15262452],
         [0.95265203, 0.51491754, 0.97241781]],

        [[0.82697679, 0.25067841, 0.53431884],
         [0.76160149, 0.57428596, 0.31079626],
         [0.38543292, 0.04917967, 0.11564118],
         ...,
         [0.85525986, 0.96466994, 0.57490453],
         [0.79782342, 0.3035675 , 0.36291843],
         [0.5218361 , 0.88650856, 0.82114362]],

        [[0.3028676 , 0.26858661, 0.8353516 ],
         [0.31205638, 0.03510538, 0.31696766],
         [0.5890409 , 0.76158601, 0.46040545],
         ...,
         [0.45226136, 0.02066657, 0.80340209],
         [0.80623406, 0.54176787, 0.79094872],
         [0.68333857, 0.94943435, 0.77650983]],

        ...,

        [[0.83541312, 0.60582576, 0.89939743],
         [0.81113274, 0.9120142 , 0.0389259 ]

#### 텐서의 속성

- 축/차원 (Rank): ndim
- 형 Shape: shape
- 데이터 타입: float32, uint8, float64 등

크기 size

In [16]:
# 축(Axis, Rank), 차원
mov.ndim

5

In [17]:
# 행
mov.shape

(10, 30, 128, 128, 3)

In [18]:
# 크기
mov.size

14745600

In [19]:
mov.dtype

dtype('float64')

### 학습 데이터

In [20]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()


In [21]:
train_images.shape, train_labels.shape, test_images.shape, test_labels.shape

((60000, 28, 28), (60000,), (10000, 28, 28), (10000,))

In [22]:
# 10:100 사이 데이터
slice = train_images[10:100]
slice.shape, slice.ndim

((90, 28, 28), 3)

In [76]:
# 10:100 사이의 28x28 데이터를 추출
slice = train_images[10:100, :28, :28]
slice.shape, slice.ndim

((90, 28, 28), 3)

In [24]:
# 28x28 이미에서 오른쪽 14x14 만 추출
slice = train_images[:, 14:, 14:]
slice.shape, slice.ndim

((60000, 14, 14), 3)

### 배치 데이터

딥러닝 모델은 한 번에 전체 데이터를 처리하지 않고 데이터를 작은 배치(batch) 단위로 나누어 처리한다.

In [25]:
#크기 128인 배치
batch = train_images[:128]

In [26]:
#다음 크기 128인 배치
batch = train_images[128:256]
#n번째 배치
n = 10
batch = train_images[128 * n:128 * (n+1)]

# 2. Constant

Constant 는 변경 불가능한 텐서를 생성해서 주로 고정된 데이터나 상수를 정의할 때 사용합니다.
 - 상수는 텐서플로우의 텐서 기반으로 생성된다.
 - https://www.tensorflow.org/api_docs/python/tf/Tensor

In [27]:
tf.constant

<function tensorflow.python.framework.constant_op.constant(value, dtype=None, shape=None, name='Const') -> Union[tensorflow.python.framework.ops.Operation, tensorflow.python.framework.ops._EagerTensorBase]>

- 상수를 정의한다.

In [28]:
# 상수 10, 20 를 정의한다.
a = tf.constant(10)
b = tf.constant(20, name="cons_a")

텐서 기반으로 랭크(차원), 형태, 데이터형식으로 구현되어 있다.

In [29]:
a

<tf.Tensor: shape=(), dtype=int32, numpy=10>

In [30]:
b

<tf.Tensor: shape=(), dtype=int32, numpy=20>

In [31]:
# 정수형 텐서 상수를 하나 선언하자.
a = tf.constant(10)
a.dtype

tf.int32

- 상수 텐서의 dtype은 변경이 불가

In [32]:
# 정수형을 tf.float32 로 변경해 보자
# a.dtype = tf.float32

### 텐서의 기본 속성


텐서 속성과 메소드를 확인해 보자

In [33]:
for i in dir(a) :
    if not i.startswith("_") :
        print(i)

OVERLOADABLE_OPERATORS
backing_device
consumers
cpu
device
dtype
eval
experimental_ref
get_shape
gpu
graph
is_packed
name
ndim
numpy
op
ref
set_shape
shape
value_index


In [34]:
a.dtype
a.device
a.backing_device
a.shape
a.ndim

0

In [35]:
a.numpy()

np.int32(10)

In [36]:
b.numpy()

np.int32(20)

### 문자열 상수

hello world 출력

In [37]:
# 상수: Hello World 값
hello = tf.constant('Hello, tensorflow !!')
print(hello)

tf.Tensor(b'Hello, tensorflow !!', shape=(), dtype=string)


### 타입 지정

타입 지정 할 경우는 실제 텐서 생성할 때 타입을 dtype에 정의해야 한다

In [38]:
# 10 을 float32 형식으로 선언
f = tf.constant(10, dtype=tf.float32)
f

<tf.Tensor: shape=(), dtype=float32, numpy=10.0>

In [39]:
f.dtype

tf.float32

In [40]:
f.device

'/job:localhost/replica:0/task:0/device:GPU:0'

### 다차원 텐서

In [41]:
# 행렬 (2차원 텐서)
matrix = tf.constant([[10, 20], [30, 40]])
print(f"Matrix: {matrix}, Shape: {matrix.shape}, Dtype: {matrix.dtype}")

Matrix: [[10 20]
 [30 40]], Shape: (2, 2), Dtype: <dtype: 'int32'>


In [42]:
matrix.dtype

tf.int32

In [43]:
matrix = tf.ones([10,10], dtype=tf.int32)
matrix

<tf.Tensor: shape=(10, 10), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=int32)>

In [44]:
matrix.dtype

tf.int32

# 3. Operation (연산)

텐서 사이의 사칙연산, 브로드캐스팅 등의 다양한 수치 연산을 수행한다
- TensorFlow는 다양한 수학 연산, 행렬 연산, 비교 연산 등을 제공하여 복잡한 계산 그래프를 구축할 수 있도록 합니다. 



### 기본 수학 연산

 - 덧셈: tf.add(a, b) 또는 a + b
 - 뺄셈: tf.subtract(a, b) 또는 a - b
 - 곱셈: tf.multiply(a, b) 또는 a * b (요소별 곱셈)
 - 나눗셈: tf.divide(a, b) 또는 a / b
 

In [45]:
a = tf.constant(10)
b = tf.constant(32)

In [46]:
# add()
c = tf.add(a,b)

# substract()
tf.subtract(a, b)

<tf.Tensor: shape=(), dtype=int32, numpy=-22>

In [47]:
# multiply
tf.multiply(a, b)

# divide()
tf.divide(a, b)

<tf.Tensor: shape=(), dtype=float64, numpy=0.3125>

In [48]:
c = tf.constant(10)
d = tf.constant(20)



In [49]:
k = tf.constant( 10)
j = 20
# divide k, j
tf.divide(k,j)

<tf.Tensor: shape=(), dtype=float64, numpy=0.5>

연산자를 사용한 연산도 가능하다

In [50]:
a + b
a - b
a * b
a / b
a % b

<tf.Tensor: shape=(), dtype=int32, numpy=10>

In [51]:
# to numpy
a.numpy()

np.int32(10)

### 반올림처리

In [52]:
# floor
tf.floor(10.5)

<tf.Tensor: shape=(), dtype=float32, numpy=10.0>

In [53]:
# round
tf.round(10.5)

<tf.Tensor: shape=(), dtype=float32, numpy=10.0>

### 절대값 및 부호처리

In [54]:
tf.abs(-100)

<tf.Tensor: shape=(), dtype=int32, numpy=100>

In [55]:
tf.sign

<function tensorflow.python.ops.math_ops.sign(x, name=None)>

In [56]:
tf.negative

<function tensorflow.python.ops.gen_math_ops.neg(x: Annotated[_any, ~TV_Neg_T], name=None) -> Annotated[_any, ~TV_Neg_T]>

### 제곱 및 제곱근, 지수

In [57]:
tf.pow

<function tensorflow.python.ops.math_ops.pow(x, y, name=None)>

In [58]:
tf.pow(10,2)

<tf.Tensor: shape=(), dtype=int32, numpy=100>

In [59]:
tf.square

<function tensorflow.python.ops.gen_math_ops.square(x: Annotated[_any, ~TV_Square_T], name=None) -> Annotated[_any, ~TV_Square_T]>

## Tensor broadcasting

### Constant 1-D Tensor populated with value list.

In [60]:
tensor1 = tf.constant([1, 2, 3, 4, 5, 6])
# 상수 10을 + 브로드캐스트
tensor1_ = tensor1 + 10
tensor1_

<tf.Tensor: shape=(6,), dtype=int32, numpy=array([11, 12, 13, 14, 15, 16], dtype=int32)>

In [61]:
tensor1_.shape, tensor1_.ndim

(TensorShape([6]), 1)

In [62]:
tensor1_.backing_device

'/job:localhost/replica:0/task:0/device:CPU:0'

In [63]:
tensor1.cpu()

Instructions for updating:
Use tf.identity with explicit device placement instead.


<tf.Tensor: shape=(6,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6], dtype=int32)>

### 특정 형상을 기준으로 동일한 값을 저장

In [69]:
# fill() 
fill_ = tf.fill([2, 3], 9)

In [70]:
fill_

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[9, 9, 9],
       [9, 9, 9]], dtype=int32)>

In [71]:
# 30을 + 브로드캐스트
fill_ +30

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[39, 39, 39],
       [39, 39, 39]], dtype=int32)>

## CPU, GPU 를 사용한 계산

> GPU 메모리 문제 발생 가능

### CPU

In [81]:
with tf.device('/CPU:' + str(0)):
    # 1. 연산에 사용할 행렬의 크기 정의
    matrix_size = 8192  # 더 큰 값일수록 CPU와 GPU 간의 성능 차이를 명확히 볼 수 있습니다.
    a = tf.random.normal([matrix_size, matrix_size], dtype=tf.float32)
    b = tf.random.normal([matrix_size, matrix_size], dtype=tf.float32)

In [83]:
import time

cpu_slot = 0

# Using CPU at slot 0
with tf.device('/CPU:' + str(cpu_slot)):
    start = time.time()

    c_cpu  = tf.matmul(a, b) # 행렬 곱셈
    r = c_cpu .numpy()

    end = time.time() - start
    print(f"CPU 연산 시간: {end:.4f} 초")


CPU 연산 시간: 2.6104 초


### GPU 를 사용한 계산

In [ ]:
# Using the GPU at slot 0
gpu_slot = 0

if tf.config.list_physical_devices('GPU'):

    with tf.device('/GPU:' + str(gpu_slot)):
        # Starting a timer
        start = time.time()

        # Doing operations on GPU
        # A = tf.constant([[3, 2], [5, 2]])
        # print(tf.eye(2,2))
        c_gpu  = tf.matmul(a, b) # 행렬 곱셈
        r = c_gpu .numpy() # 연산이 실제로 완료되도록 결과를 numpy 배열로 변환 (지연 실행 방지)

        # Printing how long it took with CPU
        end = time.time() - start
        print(f"GPU 연산 시간: {end:.4f} 초")
else:
    print("GPU를 찾을 수 없습니다. GPU 연산을 건너뜝니다.")

    

In [ ]:
gpu_slot = 0

if tf.config.list_physical_devices('GPU'):

    with tf.device('/GPU:' + str(gpu_slot)):
        start = time.time()

        c_gpu  = tf.matmul(a, b) # 행렬 곱셈
        r = c_gpu .numpy()

        end = time.time() - start
        print(f"GPU 연산 시간: {end:.4f} 초")
else:
    print("GPU를 찾을 수 없습니다. GPU 연산을 건너뜝니다.")
    

GPU 연산 시간: 0.1988 초


In [ ]:
# 5. 결과 검증
# CPU와 GPU에서 계산된 결과가 동일한지 확인합니다.
if tf.config.list_physical_devices('GPU'):
    # 두 결과가 완전히 같지 않을 수 있습니다 (부동 소수점 정밀도 차이).
    # 따라서 어느 정도의 오차 범위 내에서 같은지 확인합니다.
    diff = tf.reduce_max(tf.abs(c_cpu - c_gpu))
    print(f"\nCPU와 GPU 결과의 최대 차이: {diff:.4e}")
    if diff < 1e-5: # 작은 값으로 오차 허용 범위 설정
        print("CPU와 GPU 연산 결과가 거의 동일합니다.")
    else:
        print("CPU와 GPU 연산 결과에 차이가 있습니다.")


CPU와 GPU 결과의 최대 차이: 1.6217e-01
CPU와 GPU 연산 결과에 차이가 있습니다.


TensorFlow 예제를 실행했을 때 GPU가 CPU보다 더 느리게 작동하는 경우는 몇 가지 원인이 있을 수 있습니다. 특히 행렬 곱셈과 같은 연산에서 GPU는 일반적으로 CPU보다 훨씬 빠르다고 알려져 있는데도 말이죠.

가장 흔한 원인들을 아래에서 설명해 드리겠습니다.

1. GPU 오버헤드 (Overhead)

GPU는 병렬 처리에 특화되어 있지만, GPU 연산을 시작하고 데이터를 주고받는 데에는 초기 설정(setup) 및 데이터 전송(data transfer) 오버헤드가 발생합니다.

- 데이터 전송 (CPU to GPU, GPU to CPU): CPU 메모리에 있는 데이터를 GPU 메모리로 전송하고, 연산이 끝난 후 GPU 메모리에서 CPU 메모리로 다시 전송하는 데 시간이 소요됩니다. 이 전송 비용이 실제 GPU 연산 시간보다 더 클 수 있습니다.

- 커널 런치 (Kernel Launch) 오버헤드: CPU가 GPU에 작업을 지시(커널 런치)하는 과정에도 약간의 시간이 걸립니다.


2. GPU 워밍업 (Warm-up) 시간
일부 GPU 드라이버나 런타임 환경에서는 첫 번째 GPU 연산을 수행할 때 초기화 또는 "워밍업" 시간이 필요할 수 있습니다

# 4. 변수

TensorFlow 2.x에서는 텐서 연산에 **Eager Execution(즉시 실행)**이 기본으로 설정되어 있어, 코드를 작성하는 즉시 연산이 실행되고 결과가 반환됩니다. 이는 Python의 일반적인 코딩 방식과 유사하여 디버깅 및 개발이 훨씬 용이합니다.
> - tensorflow 2.x 이전 버전의 주요 변수 선언자: 플레이스홀더.

### 변수란?

**변수 (Variable)** 는 딥러닝 모델의 학습 가능한 파라미터(예: 가중치, 편향)를 저장하는 데 사용되는 특별한 종류의 텐서로 `tf.constant` 와 달리 `tf.Variable` 은 값을 변경할 수 있다. 모델이 데이터를 학습하면서 이 변수들의 값이 최적의 성능을 낼 수 있도록 점진적으로 업데이트된다.

### 변수 초기화

변수는 모델 학습을 시작하기 전에 적절한 값으로 초기화되어야 합니다. TensorFlow는 tf.random.normal (정규 분포), tf.zeros (모든 요소가 0), tf.ones (모든 요소가 1) 등 다양한 초기화 함수를 제공합니다.

tf.Variable() 함수를 사용하여 변수를 생성하고 초기값을 필수로 지정해야 한다.

```python
X = tf.Variable(5.0)
x_data = [[1,2,3],[4,5,6]]
X = tf.Variable(x_data)
X = tf.Variable(tf.random.normal((2,3)))
```

In [ ]:
X = tf.Variable(5.0)
X

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=5.0>

- assign 을 사용해 변수 텐서의 값을 변경할 수 있다

In [ ]:
# assign 을 사용해 변수 텐서의 값을 변경할 수 있다
X.assign(9.0)
X

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=9.0>

In [ ]:
x_data = [[1,2,3],[4,5,6]]
X = tf.Variable(x_data)
X

<tf.Variable 'Variable:0' shape=(2, 3) dtype=int32, numpy=
array([[1, 2, 3],
       [4, 5, 6]], dtype=int32)>

tf.random.normal() 등의 임의의 범위 값으로 초기화를 할 수 있다.

In [ ]:
X = tf.Variable(tf.random.normal((2,3)))
X


<tf.Variable 'Variable:0' shape=(2, 3) dtype=float32, numpy=
array([[ 1.3458368 ,  0.3061433 , -0.01974057],
       [-0.28733382, -1.1497934 ,  0.16760664]], dtype=float32)>

In [ ]:
X.assign_add(tf.random.normal((2,3)))
X

<tf.Variable 'Variable:0' shape=(2, 3) dtype=float32, numpy=
array([[ 2.9920444 ,  0.00541398,  0.83918726],
       [-0.08768904, -2.4048605 , -1.0975332 ]], dtype=float32)>

### 변수를 사용해서 방정식의 해를 찾아 보자

In [ ]:
X = tf.Variable(tf.random.normal((2,3)))

In [ ]:
W = tf.Variable(tf.random.normal((3,2)))

In [ ]:
b = tf.Variable(tf.random.normal((2,1)))

In [ ]:
expr = tf.matmul(X, W) + b

In [ ]:
expr

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[-0.16000828, -0.71595764],
       [ 0.04002523,  4.73197   ]], dtype=float32)>

In [ ]:
expr.numpy()

array([[-0.16000828, -0.71595764],
       [ 0.04002523,  4.73197   ]], dtype=float32)

### 실습: 변수와 concat 함수 사용

In [ ]:
# Making a constant tensor A, that does not change
A = tf.constant([[3, 2],
                 [5, 2]])

# Making a Variable tensor VA, which can change. Notice it's .Variable
VA = tf.Variable([[3, 2],
                 [5, 2]])

# Making another tensor B
B = tf.constant([[9, 5],
                 [1, 3]])

In [ ]:
# Concatenate columns
AB_concatenated = tf.concat(values=[A, B], axis=1)

print(('Adding B\'s columns to A:\n{0}').format( AB_concatenated.numpy()))

Adding B's columns to A:
[[3 2 9 5]
 [5 2 1 3]]


In [ ]:
# Concatenate rows
AB_concatenated = tf.concat(values=[A, B], axis=0)

print(('\nAdding B\'s rows to A:\n{0}').format( AB_concatenated.numpy()))


Adding B's rows to A:
[[3 2]
 [5 2]
 [9 5]
 [1 3]]


### reshape

In [ ]:
# Making a tensor for reshaping
tensor = tf.constant([[3, 2],
                      [5, 2],
                      [9, 5],
                      [1, 3]])

In [ ]:
tensor

<tf.Tensor: shape=(4, 2), dtype=int32, numpy=
array([[3, 2],
       [5, 2],
       [9, 5],
       [1, 3]], dtype=int32)>

In [ ]:
# Reshaping the tensor into a shape of: shape = [rows, columns]
reshaped_tensor = tf.reshape(tensor = tensor,
                             shape = [1, 8])

print(('Tensor BEFORE reshape:\n{0}').format(tensor.numpy()))
print(('\nTensor AFTER reshape:\n{0}').format(reshaped_tensor.numpy()))

Tensor BEFORE reshape:
[[3 2]
 [5 2]
 [9 5]
 [1 3]]

Tensor AFTER reshape:
[[3 2 5 2 9 5 1 3]]


### 변수형 변경: cast

In [ ]:
# Making a tensor
tensor = tf.constant([[3.1, 2.8],
                      [5.2, 2.3],
                      [9.7, 5.5],
                      [1.1, 3.4]],
                      dtype=tf.float32)

tensor_as_int = tf.cast(tensor, tf.int32)

print(('Tensor with floats:\n{0}').format(tensor.numpy()))
print(('\nTensor cast from float to int (just remove the decimal, no rounding):\n{0}').format(tensor_as_int.numpy()))

Tensor with floats:
[[3.1 2.8]
 [5.2 2.3]
 [9.7 5.5]
 [1.1 3.4]]

Tensor cast from float to int (just remove the decimal, no rounding):
[[3 2]
 [5 2]
 [9 5]
 [1 3]]


### (참고) TensorFlow 1.x과 2.0의 Eager Execution

TensorFlow 1.x의 Graph 와 Session 은 TensorFlow 2.x에서는 Eager Execution으로 지원하고 있다.

- TensorFlow 1.x에서는 모든 연산이 **그래프 (Graph)**라는 정적 계산 그래프를 먼저 정의한 후, 세션 (Session) 내에서 실행되는 방식이었습니다. 즉, "모델을 먼저 설계하고, 나중에 데이터를 넣어 실행"하는 방식이었다.

  - 그래프: 연산 노드(Operation)와 텐서 노드(Tensor)로 구성된 일종의 설계도입니다.
  - 세션: 그래프를 실행하고 실제 연산을 수행하는 런타임 환경입니다.

TensorFlow 2.x에서는 **Eager Execution (즉시 실행)**이 기본이 되면서 이러한 그래프-세션 개념이 추상화되고 사용자가 직접적으로 다룰 필요가 없어졌습니다. 코드를 작성하는 즉시 연산이 실행되어 마치 일반적인 Python 코드처럼 느껴집니다. 이는 디버깅을 훨씬 쉽게 만들고 개발 속도를 높여줍니다. 하지만 내부적으로는 여전히 tf.function과 같은 기능을 통해 그래프 최적화가 이루어지고 있습니다.


- 플래이스 홀더와 세션이 사라짐

In [ ]:
try :
    tf.placeholder('float')

except Exception as e :
    print(e)

module 'tensorflow' has no attribute 'placeholder'


In [ ]:
try :
    tf.Session()
except Exception as e :
    print(e)

module 'tensorflow' has no attribute 'Session'


In [ ]:
if(tf.executing_eagerly()):
    print('Eager execution is enabled (running operations immediately)\n')
    print(('Turn eager execution off by running: \n{0}\n{1}').format('' \
        'from tensorflow.python.framework.ops import disable_eager_execution', \
        'disable_eager_execution()'))
else:
    print('You are not running eager execution. TensorFlow version >= 2.0.0' \
          'has eager execution enabled by default.')
    print(('Turn on eager execution by running: \n\n{0}\n\nOr upgrade '\
           'your tensorflow version by running:\n\n{1}').format(
           'tf.compat.v1.enable_eager_execution()',
           '!pip install --upgrade tensorflow\n' \
           '!pip install --upgrade tensorflow-gpu'))

Eager execution is enabled (running operations immediately)

Turn eager execution off by running: 
from tensorflow.python.framework.ops import disable_eager_execution
disable_eager_execution()


# 4. 행렬

- 영행렬 일행렬
- 채움
- 전치행렬
- 대각행렬
- 항등행렬

### 영행렬, 일행렬
 -  zeros()
 -  ones()

In [ ]:
# Making a tensor filled with zeros. shape=[rows, columns]
tensor = tf.zeros(shape=[3, 4], dtype=tf.int32)
print(('Tensor full of zeros as int32, 3 rows and 4 columns:\n{0}').format(tensor.numpy()))

Tensor full of zeros as int32, 3 rows and 4 columns:
[[0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]]


In [ ]:
# Making a tensor filled with zeros with data type of float32
tensor = tf.ones(shape=[5, 3], dtype=tf.float32)
print(('\nTensor full of ones as float32, 5 rows and 3 columns:\n{0}').format(tensor.numpy()))


Tensor full of ones as float32, 5 rows and 3 columns:
[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


### fill()

In [ ]:
tensor_fill = tf.fill([10,10],0)
tensor_fill

<tf.Tensor: shape=(10, 10), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)>

In [ ]:
tf.ones([10,10])

<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)>

### 전치행렬
Transpose

In [ ]:
# Some Matrix A
A = tf.constant([[3, 7],
                 [1, 9]])

A = tf.transpose(A)

print(('The transposed matrix A:\n{0}').format(A))

The transposed matrix A:
[[3 1]
 [7 9]]


In [ ]:
E = tf.constant([[1, 2, 3], [4, 5, 6]])
E

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[1, 2, 3],
       [4, 5, 6]], dtype=int32)>

In [ ]:
# Transposing Matrix E
transpose_E = tf.transpose(E, name="transposeE")
transpose_E

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 4],
       [2, 5],
       [3, 6]], dtype=int32)>

### 대각행렬

In [ ]:
# 주 대각선 원소 정의
diagonal_elements = tf.constant([1, 2, 3])

# 대각 행렬 생성
diagonal_matrix = tf.linalg.diag(diagonal_elements)

print("대각 행렬:")
print(diagonal_matrix.numpy())

대각 행렬:
[[1 0 0]
 [0 2 0]
 [0 0 3]]


### Identity Matrix

항등행렬

In [ ]:
# Some Matrix A
A = tf.constant([[3, 7],
                 [1, 9],
                 [2, 5]])

# Get number of dimensions
rows, columns = A.shape
print(('in tensor A:\n{0} rows\n{1} columns').format(rows, columns))

# Making identity matrix
A_identity = tf.eye(num_rows = rows,
                    num_columns = columns,
                    dtype = tf.int32)
print(('\nidentity matrix of A:\n{0}').format(
    A_identity.numpy()))

in tensor A:
3 rows
2 columns

identity matrix of A:
[[1 0]
 [0 1]
 [0 0]]


## 행렬의 연산

- 브로드캐스팅
- 행렬 곱셈: tf.matmul(a, b) 또는 a @ b 

### 브로드캐스팅

Element-wise broadcasting 은 크기가 다른 텐서 간의 연산을 가능하게 하는 메커니즘으로 NumPy의 브로드캐스팅 규칙과 유사하다. 작은 텐서가 큰 텐서의 형태에 맞춰 자동으로 확장되어 연산이 수행된다.

 - NumPy, Keras, Theano, Tensorflow 의 원소별 곱셈을 `*` 연산자를 사용한다.


In [ ]:
A = tf.constant([[3, 2],
                 [5, 2]])
A * 5
A + 5

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 8,  7],
       [10,  7]], dtype=int32)>

multiply 등 연산 함수를 이용할 수 있다.

In [ ]:
# Element-wise multiplication
Av = tf.multiply(A, 5)

In [ ]:
# Element-wise multiplication
v = tf.constant([[5],
                 [2]])
Av = tf.multiply(A, v)
Av

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[15, 10],
       [10,  4]], dtype=int32)>

### 행렬 내적


- 행렬의 내적(dot product)은 tf.matmul(a, b) 또는 a @ b 를 지원 

> - dot product: 주로 벡터 (1차원) 간의 연산이며, 결과는 스칼라입니다.
> - 행렬의 곱셈: 
>    - tf.multiply()
>    - 행렬 (2차원 이상) 간의 연산이며, 결과는 행렬입니다



  - TensorFlow (및 NumPy)에서 `tf.tensordot` 같은 함수를 사용하거나, 1차원 벡터에 대해 `tf.matmul`을 사용할 경우 내적의 의미로 작동할 수 있습니다. 



Dot produt (여기서 행렬곱) 계산은
 - NumPy에서 벡터와 행렬의 내적은 `np.dot()` 함수사용.
 - Python3.5 이상은 `@`(at이라고 읽는다) 연산자로 계산한다.
     - 2차원 배열로 표시한 벡터를 내적했을 때는 결과값이 스칼라가 아닌 2차원 배열이다.
 - Keras는 keras.dot() 연산 사용.
 - 텐서플로우는 tf.matmul(x, y) 를 사용한다.
 
<img src='https://i.imgur.com/TahVEIY.png' width=500>

https://angeloyeo.github.io/2020/09/08/matrix_multiplication.html

**행렬 곱셈 (Matrix Multiplication)**은 수학적으로 정의된 행렬 간의 곱셈 규칙을 따르는 연산입니다. 딥러닝에서 신경망의 레이어 간 계산(입력 벡터와 가중치 행렬의 곱 등)에 가장 많이 사용됩니다.


In [ ]:
matrix1 = tf.constant([[1, 2], [3, 4]])
matrix2 = tf.constant([[5, 6], [7, 8]])

matrix_mult = tf.matmul(matrix1, matrix2)
print(f"Matrix multiplication:\n{matrix_mult.numpy()}")

Matrix multiplication:
[[19 22]
 [43 50]]


# 5. tf.function 그래프 컴파일


TensorFlow 2.x의 Eager Execution은 개발 편의성을 높였지만, 파이썬 코드의 인터프리터 오버헤드 때문에 순수 C++로 구현된 TensorFlow 1.x의 정적 그래프 방식보다 속도가 느릴 수 있습니다. **tf.function**은 이 문제를 해결하기 위해 사용됩니다.


- https://www.tensorflow.org/guide/intro_to_graphs?hl=ko


#### - `tf.function`의 역할

 - Python 함수를 호출 가능한 \*\*TensorFlow 그래프(TensorFlow Graph)\*\*로 컴파일합니다.
 - 이렇게 컴파일된 그래프는 Python 인터프리터의 간섭 없이 TensorFlow 런타임에서 직접 실행되므로 **성능이 크게 향상**됩니다.
 - GPU나 TPU 같은 가속기에서도 더 효율적으로 동작합니다.
 - 모델을 배포할 때 (SavedModel)에도 `tf.function`으로 캡슐화된 코드는 그래프 형태로 저장되어 쉽게 로드하고 실행할 수 있습니다.


## **`tf.function` 사용**:

  -  파이썬 함수 위에 데코레이터로 `@tf.function`을 붙이면 됩니다.
  - tf.function() 으로 함수를 전달한다
  - 첫 호출 시 함수는 그래프로 변환됩니다. 이후 호출부터는 변환된 그래프가 실행됩니다.


In [ ]:
import tensorflow as tf
import time

#### - `tf.function` 즉시 사용

In [ ]:
def rms(x, y):
    return tf.reduce_mean(tf.multiply(x ** 2, 3) + y)

r = tf.function(rms)


In [ ]:
r

In [ ]:
x = tf.constant([[2.0, 3.0]])
y = tf.constant([[3.0, -2.0]])

# `rms()` and `r` will return the same value,
# but `r` will be executed as a TensorFlow graph.
assert rms(x, y).numpy() == r(x, y).numpy()

print(rms(x,y).numpy())

20.0


#### - 데코레이터 사용

In [ ]:
@tf.function
def rms2(x, y) :
    return tf.reduce_mean(tf.multiply(x ** 2, 3) + y)

In [ ]:
x = tf.constant([[2.0, 3.0]])
y = tf.constant([[3.0, -2.0]])

s = rms2(x, y)
s

<tf.Tensor: shape=(), dtype=float32, numpy=20.0>

In [ ]:
# `rms()` and `r` will return the same value,
# but `r` will be executed as a TensorFlow graph.
assert rms(x, y).numpy() == rms2(x, y).numpy()

print(rms(x,y).numpy())

20.0


#### - 성능 비교

In [ ]:
# 일반 Python 함수
def python_add(a, b):
    return a + b

# tf.function으로 컴파일된 함수
@tf.function
def tf_function_add(a, b):
    return a + b

In [ ]:
# 큰 텐서 생성
tensor_a = tf.random.normal([10000, 10000])
tensor_b = tf.random.normal([10000, 10000])

print("\n--- tf.function 성능 비교 ---")

# Python 함수 실행 시간 측정
start_time = time.time()
for _ in range(100):
    result = python_add(tensor_a, tensor_b)
end_time = time.time()
print(f"Python function execution time: {end_time - start_time:.4f} seconds")

# tf.function 함수 실행 시간 측정 (첫 호출은 컴파일 포함)
start_time = time.time()
for _ in range(100):
    result = tf_function_add(tensor_a, tensor_b)
end_time = time.time()
print(f"tf.function execution time (including compilation): {end_time - start_time:.4f} seconds")

# tf.function 함수 재실행 시간 측정 (컴파일 완료 후)
start_time = time.time()
for _ in range(100):
    result = tf_function_add(tensor_a, tensor_b) # 이미 컴파일됨
end_time = time.time()
print(f"tf.function execution time (after compilation): {end_time - start_time:.4f} seconds")



--- tf.function 성능 비교 ---
Python function execution time: 0.0125 seconds
tf.function execution time (including compilation): 0.0960 seconds
tf.function execution time (after compilation): 0.0646 seconds


In [ ]:
def rms(x, y) :
    return tf.reduce_mean(tf.multiply(x ** 2, 3) + y)

In [ ]:
@tf.function
def tf_rms(x, y) :
    return tf.reduce_mean(tf.multiply(x ** 2, 3) + y)

In [ ]:
# 큰 텐서 생성
tensor_a = tf.random.normal([10000, 10000])
tensor_b = tf.random.normal([10000, 10000])

print("\n--- tf.function 성능 비교 ---")

# Python 함수 실행 시간 측정
start_time = time.time()
for _ in range(100):
    result = rms(tensor_a, tensor_b)
end_time = time.time()
print(f"Python function execution time: {end_time - start_time:.4f} seconds")

# tf.function 함수 실행 시간 측정 (첫 호출은 컴파일 포함)
start_time = time.time()
for _ in range(100):
    result = tf_rms(tensor_a, tensor_b)
end_time = time.time()
print(f"tf.function execution time (including compilation): {end_time - start_time:.4f} seconds")

# tf.function 함수 재실행 시간 측정 (컴파일 완료 후)
start_time = time.time()
for _ in range(100):
    result = tf_rms(tensor_a, tensor_b) # 이미 컴파일됨
end_time = time.time()
print(f"tf.function execution time (after compilation): {end_time - start_time:.4f} seconds")



--- tf.function 성능 비교 ---
Python function execution time: 0.0661 seconds
tf.function execution time (including compilation): 2.9240 seconds
tf.function execution time (after compilation): 0.0697 seconds


이 예시를 실행해보면, `tf.function`으로 캡슐화된 함수가 훨씬 빠르게 동작하는 것을 확인할 수 있습니다. 특히 루프 내에서 반복적으로 호출되거나 복잡한 연산을 수행할 때 그 효과가 두드러집니다.


### **`tf.function` 사용 시 주의사항**

  - `tf.function` 내부에서는 Python의 일반적인 제어 흐름(for 루프, if/else 등)이 TensorFlow 연산으로 변환될 수 있도록 작성해야 합니다. `tf.Tensor` 객체를 이용한 조건문이나 반복문은 잘 동작하지만, 일반 Python 리스트나 넘파이 배열에 대한 직접적인 조작은 그래프로 변환되지 않을 수 있습니다.
  - 변수가 아닌 Python 원시 타입(정수, 문자열)이 함수의 입력으로 들어오면, 입력값마다 별도의 그래프를 생성할 수 있어 비효율적일 수 있습니다. 텐서 형태의 입력을 사용하는 것이 권장됩니다.


## Tensorflow 주요 함수

#### 논리연산

In [ ]:
tf.logical_and
tf.logical_or
tf.logical_not

<function tensorflow.python.ops.gen_math_ops.logical_not(x: Annotated[_any, <class 'tensorflow.security.fuzzing.py.annotation_types.Bool'>], name=None) -> Annotated[_any, <class 'tensorflow.security.fuzzing.py.annotation_types.Bool'>]>

#### 비교연산

In [ ]:
tf.equal
tf.less
tf.less_equal
tf.greater
tf.greater_equal

<function tensorflow.python.ops.gen_math_ops.greater_equal(x: Annotated[_any, ~TV_GreaterEqual_T], y: Annotated[_any, ~TV_GreaterEqual_T], name=None) -> Annotated[_any, <class 'tensorflow.security.fuzzing.py.annotation_types.Bool'>]>

#### 최대값 최소값

In [ ]:
tf.maximum
tf.minimum

<function tensorflow.python.ops.gen_math_ops.minimum(x: Annotated[_any, ~TV_Minimum_T], y: Annotated[_any, ~TV_Minimum_T], name=None) -> Annotated[_any, ~TV_Minimum_T]>

#### 삼각함수

In [ ]:
tf.cos
tf.sin
tf.tan
tf.cosh
tf.sinh
tf.tanh

<function tensorflow.python.ops.gen_math_ops.tanh(x: Annotated[_any, ~TV_Tanh_T], name=None) -> Annotated[_any, ~TV_Tanh_T]>

#### argmax, argmin

1. [argmax](https://www.tensorflow.org/api_docs/python/tf/argmax)
2. [argmin](https://www.tensorflow.org/api_docs/python/tf/argmin)




In [ ]:
x = [-10, 9, 8, 10]
tf.argmax(x)
tf.argmin(x)

<tf.Tensor: shape=(), dtype=int64, numpy=0>

## one_hot()

[one_hot()](https://www.tensorflow.org/api_docs/python/tf/one_hot)

one-hot(원핫)인코딩이란? 

원-핫 인코딩은 특정 위치의 값만 1이고, 나머지 위치의 값은 0인 벡터를 만드는 방식입니다.

예를들면 `[0, 0, 0, 0, 1]` 이다. 5번째만 1이고 나머지는 0이다. 



In [ ]:
tf.one_hot([0,1,2], depth=3)

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]], dtype=float32)>

In [ ]:
tf.one_hot([0,4,2], depth=3)

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[1., 0., 0.],
       [0., 0., 0.],
       [0., 0., 1.]], dtype=float32)>

In [ ]:
tf.one_hot([1,2,3], depth=5)

<tf.Tensor: shape=(3, 5), dtype=float32, numpy=
array([[0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.]], dtype=float32)>

## random 모듈

분포 함수
 - 정규분포에 대한 임의의 값 세팅

In [ ]:
tf.random.normal((2,3))
tf.random.uniform([4],0,1)
tf.random.uniform([4,2],0,1).numpy()
tf.random.uniform([4,2,2],0,1).numpy()

array([[[0.37055755, 0.17974937],
        [0.10512745, 0.6269852 ]],

       [[0.3236947 , 0.27060676],
        [0.8151705 , 0.14533818]],

       [[0.10854578, 0.09838355],
        [0.7482178 , 0.15536332]],

       [[0.3580339 , 0.0248915 ],
        [0.75883186, 0.64871335]]], dtype=float32)

## 배열 변형 함수

- [reshape()](https://www.tensorflow.org/api_docs/python/tf/reshape
): 원하는 shape를 직접 입력하여 바꿀 수 있다. 특히 shape에 -1를 입력하면 고정된 차원은 우선 채우고 남은 부분을 알아서 채워준다.
- [squeeze()](https://www.tensorflow.org/api_docs/python/tf/squeeze) 차원 중 사이즈가 1인 것을 찾아 스칼라값으로 바꿔 해당 차원을 제거한다.
- [expand_dims](https://www.tensorflow.org/api_docs/python/tf/expand_dims): axis로 지정된 차원을 추가한다.









### `reshape()`

In [ ]:
x = np.arange(0,10)
x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
tf.reshape(x, shape=[-1,2])

<tf.Tensor: shape=(5, 2), dtype=int64, numpy=
array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])>

In [ ]:
tf.reshape(x, shape=[-1,5])

<tf.Tensor: shape=(2, 5), dtype=int64, numpy=
array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])>

### `squeeze()`

In [ ]:
x = np.arange(0,10).reshape(-1,1)
x

array([[0],
       [1],
       [2],
       [3],
       [4],
       [5],
       [6],
       [7],
       [8],
       [9]])

In [ ]:
x.shape

(10, 1)

차원중 크기가 1인 것을 찾아 짜낸다.

In [ ]:
tf.squeeze(x)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])>

In [ ]:
tf.squeeze(x).shape

TensorShape([10])

### `expand_dims()`

axis로 지정된 차원을 추가한다.

In [ ]:
x = np.arange(0,10).reshape(-1,)
x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
x.shape

(10,)

In [ ]:
# 첫번째 차원을 추가
y = tf.expand_dims(x, 0)
y

<tf.Tensor: shape=(1, 10), dtype=int64, numpy=array([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]])>

In [ ]:
# 두번째 차원을 추가
y = tf.expand_dims(x, 1)
y

<tf.Tensor: shape=(10, 1), dtype=int64, numpy=
array([[0],
       [1],
       [2],
       [3],
       [4],
       [5],
       [6],
       [7],
       [8],
       [9]])>

In [ ]:
x = np.arange(0,10).reshape(-1,2)
print(x, x.shape)

[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]] (5, 2)


In [ ]:
y = tf.expand_dims(x, 0)
y

<tf.Tensor: shape=(1, 5, 2), dtype=int64, numpy=
array([[[0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]]])>

In [ ]:
y = tf.expand_dims(x, 2)
y

<tf.Tensor: shape=(5, 2, 1), dtype=int64, numpy=
array([[[0],
        [1]],

       [[2],
        [3]],

       [[4],
        [5]],

       [[6],
        [7]],

       [[8],
        [9]]])>

In [ ]:
# 마지막 차수에 차원을 추가한다.
y = tf.expand_dims(x, -1)
y

<tf.Tensor: shape=(5, 2, 1), dtype=int64, numpy=
array([[[0],
        [1]],

       [[2],
        [3]],

       [[4],
        [5]],

       [[6],
        [7]],

       [[8],
        [9]]])>

## stack()

[stack()](https://www.tensorflow.org/api_docs/python/tf/stack) 함수

 - 여러 조각의 행렬을 합칠 때 사용하는 stack함수에 대해서 알아보자.


In [ ]:
x1 = np.array([1,2])
x2 = np.array([2,3])
x3 = np.array([4,5])

In [ ]:
# stack
tf.stack([x1,x2,x3])

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 2],
       [2, 3],
       [4, 5]], dtype=int32)>

In [ ]:
# stack axis=0
tf.stack([x1,x2,x3], axis=0)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 2],
       [2, 3],
       [4, 5]], dtype=int32)>

In [ ]:
# stack axis=1
tf.stack([x1,x2,x3], axis=1)

<tf.Tensor: shape=(2, 3), dtype=int64, numpy=
array([[1, 2, 4],
       [2, 3, 5]])>

## reduce_sum, reduce_mean

2차원 이상의 행렬로 된 데이터를 다루다보면 전체 원수의 합계, 평균을 구하고자 할 때 사용한다.

1. [reduce_sum()](https://www.tensorflow.org/api_docs/python/tf/reduce_sum) 은 특정 차원을 제거하고 합계를 구하고
1. [reduce_mean()](https://www.tensorflow.org/api_docs/python/tf/reduce_mean)은 특정 차원을 제거하고 평균을 구한다.

In [ ]:
# reduce_sum
x = np.arange(0,16).reshape(-1,4)
x

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [ ]:
tf.rank(x) # 차원수

<tf.Tensor: shape=(), dtype=int32, numpy=2>

In [ ]:
tf.reduce_sum(x) # all elements

<tf.Tensor: shape=(), dtype=int64, numpy=120>

In [ ]:
tf.reduce_sum(x, axis=0) # row

<tf.Tensor: shape=(4,), dtype=int64, numpy=array([24, 28, 32, 36])>

In [ ]:
tf.reduce_sum(x, axis=1) # column

<tf.Tensor: shape=(4,), dtype=int64, numpy=array([ 6, 22, 38, 54])>

`keep_dims=True` 최초 차원수를 유지

In [ ]:
tf.reduce_sum(x, axis=1, keepdims=True) # column

<tf.Tensor: shape=(4, 1), dtype=int64, numpy=
array([[ 6],
       [22],
       [38],
       [54]])>

In [ ]:
x = np.array([[1,1,1],
              [1,1,1]])
x

array([[1, 1, 1],
       [1, 1, 1]])

In [ ]:
x.shape

(2, 3)

In [ ]:
tf.reduce_sum(x, axis=1)

<tf.Tensor: shape=(2,), dtype=int64, numpy=array([3, 3])>

In [ ]:
tf.reduce_sum(x, axis=-1) # -1번째 차원 제거

<tf.Tensor: shape=(2,), dtype=int64, numpy=array([3, 3])>

In [ ]:
tf.reduce_sum(x, axis=-2) # -1번째 차원 제거

<tf.Tensor: shape=(3,), dtype=int64, numpy=array([2, 2, 2])>